### DISCUSSION TITLE
Network with RNDIS, CDC-ECM

### DISCUSSION THREAD

**Discussion by torstenmay** [hathach/tinyusb#742] [Q&A] [answered]
Hi,

I'm thinking about using an ESP32-S2 and wondering what this statement really means:
- **Network with RNDIS, CDC-ECM (work in progress)**
- it can be found on the main Tinyusb website

Does it mean all network over USB class implementations are work in progress or just the CDC-ECM?
Is anyone using the drivers already?
Is there an overview what might be missing?

Thanks Torsten

**Comment by hathach** [comment, marked as answer]
The whole net driver is WIP, which mean API can change significantly later on if needed. You can try out the example here https://github.com/hathach/tinyusb/tree/master/examples/device/net_lwip_webserver 

But it doesn't have freeRTOS to run with esp32s2. related article on rp2040 https://www.hackster.io/news/hermann-stamm-wilbrandt-gives-the-raspberry-pi-pico-web-server-client-over-usb-capabilities-74104c572db1

**Reply by torstenmay** [reply]
Hi hathatch,

looks good.

When it's WIP what are the next steps?
- FreeRtos support?
- Https support?

Are you working on it or may I help a bit?

Cheers Torsten

**Reply by hathach** [reply]
freeRTOS already supported by the stack, https is not part of stack. WIP is mature API.

**Reply by torstenmay** [reply]
Hi hathach,

I managed to get the net_lwip_webserver example running on a esp32-s2 over the weekend.

Are you interested to discuss the solution and how it could be integrated into the Tinyusb repo?

I was forced to hack the example in some ugly way as the esp-idf come with a complete different project structure. 
Hence it would be great to get your feedback how certain adaptations should be changed that it fits into the original design.

To make some examples:
- cmake CmakeLists.txt changes
- esp-idf dependencies
- how to include the esp-idf lwipopts.h file so that precompile defines are not mixed
- main.c queue handling - When you check the queue receive func - the FreeRtos version is blocking endlessly. Hence the example can not work that way.

Regards Torsten

**Reply by hathach** [reply]
@torstenmay that is great to hear, you could follow the CMAKEList.txt examples of cdc_msc_freertos / hid_composite_freertos. All of these end with `_freertos` can be compiled and run with esp32s2 (just run `make BOARD=espressif_saola_1` with all then flash target, It should just work. You need to set the OS to freertos, for stack to properly use the correct freeRTOS queue API without hacking.  

```C
#define CFG_TUSB_OS               OPT_OS_FREERTOS
```

**Reply by torstenmay** [reply]
Hi @hathach 

i checked the two example projects and adapted some of the really ugly hacks. :)

However the network example has much more dependencies like esp_netif and lwip.

Just to give you some ideas what I have changed please have a lool here: [esp32s2.diff.txt](https://github.com/hathach/tinyusb/files/6379419/esp32s2.diff.txt)
I enabled to lwip prints and attached to output here: [log.txt](https://github.com/hathach/tinyusb/files/6379453/log.txt)

The result is that a new network interface will show up:
![image](https://user-images.githubusercontent.com/1299533/116135712-b893c600-a6d1-11eb-96ab-991c50230ece.png)

Please have a look into the patch and give me your opinion about my modifications.

Regards Torsten

**Reply by hathach** [reply]
The diff txt is not too useful for discusion, for code discuss, PR is the best way to do. Could you make an draft PR with your webserver example changes, it is not meant to be merged, but we could use the github file changes for suggestion/discusion.

**Reply by torstenmay** [reply]
Hi @hathach 

I opened the draft PR:
https://github.com/hathach/tinyusb/pull/809

Hope i did correctly as this is my first draft PR and still learning how to do it.

Regards Torsten

**Reply by hathach** [reply]
Thank you for the PR. I will check it out whenever I could.

**Comment by chegewara** [comment]
Hi,
i know it is kind of old topic, but i have question.
Espressif is having cdc-ncm example that lats us running net_lwip_webserver like code here:
https://github.com/espressif/esp-idf/tree/master/examples/network/sta2eth


My question is, because it is not clear to me, which stack setup ethernet speed detected by OS?
- is it tinyusb
- is it lwip

Im asking, because esp32P4 has USB HS and ethernet speed detected on my ubuntu is only 10Mbits.

Thanks

**Reply by HiFiPhile** [reply]
If you use NCM driver it's defined here https://github.com/hathach/tinyusb/blob/eda3cceab207bce390b995462e313ed124fa7732/src/class/net/ncm_device.c#L167
But in any case the reported speed is just an indication, won't affect the real speed at all.

**Reply by chegewara** [reply]
Actually im not asking about USB port speed (FS vs HS). I am asking about ethernet device speed, which is 10, 100, 1000Mbits/s.

Thanks

**Reply by HiFiPhile** [reply]
Yes I know you are talking about Ethernet speed.
With NCM driver reported Ethernet speed change according to USB speed.

With older ECM driver it's defined here:
https://github.com/hathach/tinyusb/blob/eda3cceab207bce390b995462e313ed124fa7732/src/class/net/ecm_rndis_device.c#L86

**Reply by chegewara** [reply]
Thank you very much. It is exactly what i am looking for, just espressif component seems to be outdated and is using one speed (10Mb)
```
tu_static struct ecm_notify_struct ncm_notify_speed_change =
{
    .header = {
        .bmRequestType_bit = {
          .recipient = TUSB_REQ_RCPT_INTERFACE,
          .type      = TUSB_REQ_TYPE_CLASS,
          .direction = TUSB_DIR_IN
        },
        .bRequest = CDC_NOTIF_CONNECTION_SPEED_CHANGE,
        .wLength = 8,
    },
    .downlink = 10000000,
    .uplink = 10000000,
};
```
https://github.com/espressif/tinyusb/blob/33d36cb23d6f8415d0e8a09d15a5e6cd8c0a6d05/src/class/net/ncm_device.c#L223-L236

Thanks
